File: src/Aerie.Api/Modules/Hatch/Runners.cs
using Microsoft.Extensions.Options;

namespace Aerie.Api.Modules.Hatch;

/// <summary>
/// The runner, and the one place its two horizons live: how long a row may go
/// unheard from before it reads <em>gone</em>, how much longer before it stops
/// being returned at all, and what a row plus whatever claim it holds looks
/// like on the wire.
///
/// A class rather than a static for the reason <see cref="IssueClaims"/> is
/// one: it holds a configured number and nothing else, and every judgement it
/// makes takes the instant it is judging against rather than reading a clock.
/// </summary>
/// <remarks>
/// Nothing here sweeps and nothing here writes. A runner that stopped answering
/// at midnight is idle at 00:01, gone at 00:02 and not in the answer at all by
/// 00:16 - each of those by arithmetic against <c>LastSeenAt</c> at the moment
/// somebody asks, which is the same lazy expiry the claim uses. There is no
/// timeout to tune and no background job to notice has stopped.
/// </remarks>
public sealed class Runners(IOptions<HatchOptions> options)
{
    /// <summary>
    /// How much longer than <see cref="GoneAfterSeconds"/> a row survives
    /// before it is dropped from the read: ten times, so the shipped default is
    /// gone after a minute and a half and forgotten after fifteen minutes.
    /// </summary>
    /// <remarks>
    /// A constant and not a setting. The two horizons are one judgement seen
    /// twice - "not answering" and "not coming back" - and an installation able
    /// to set them independently could set the second shorter than the first,
    /// which would make a row vanish before it was ever drawn as gone.
    /// </remarks>
    public const int DropMultiple = 10;

    /// <summary>
    /// The silence after which a runner reads as gone, in seconds. Guarded here
    /// and nowhere else: a configured zero or a negative would make every
    /// runner gone on arrival, and a misconfigured horizon should cost a
    /// fallback rather than a page that says every loop is dead.
    /// </summary>
    public int GoneAfterSeconds { get; } =
        options.Value.RunnerGoneAfterSeconds > 0
            ? options.Value.RunnerGoneAfterSeconds
            : new HatchOptions().RunnerGoneAfterSeconds;

    /// <summary>The heartbeat at or after which a runner is still being heard from.</summary>
    public DateTimeOffset Cutoff(DateTimeOffset now) => now.AddSeconds(-GoneAfterSeconds);

    /// <summary>The heartbeat before which a row is not handed out at all.</summary>
    public DateTimeOffset DropBefore(DateTimeOffset now) => now.AddSeconds(-GoneAfterSeconds * DropMultiple);

    /// <summary>
    /// Whether this runner is still answering. Exactly at the cutoff is here;
    /// one tick past it is gone - the same edge the claim draws, so the two
    /// never disagree about which side of a boundary an instant is on.
    /// </summary>
    public bool IsHere(DateTimeOffset lastSeenAt, DateTimeOffset now) => lastSeenAt >= Cutoff(now);

    /// <summary>
    /// Whether this row has aged out of the read. Not a deletion: the row stays
    /// where it is, and a runner that comes back under the same name picks its
    /// own bounds up again rather than being seeded afresh.
    /// </summary>
    public bool IsDropped(DateTimeOffset lastSeenAt, DateTimeOffset now) => lastSeenAt < DropBefore(now);

    /// <summary>
    /// The row as a client draws it, with whatever it holds folded in.
    /// </summary>
    /// <param name="claim">
    /// The live claim this runner holds, or null between increments. The
    /// liveness is the caller's to decide, because a scan judges every row
    /// against one instant - see <see cref="IssueClaims.IsLive"/>.
    /// </param>
    /// <remarks>
    /// The line is the claim's while there is one and the row's when there is
    /// not, and that is the whole of why the row has a line at all: an
    /// increment's chatter already rides its lease, so this column only ever
    /// carries what a runner said between tickets.
    /// </remarks>
    public RunnerDto Project(EfHatchRunner runner, (string Key, ClaimSnapshot Claim)? claim) => new(
        runner.Name,
        runner.Kind,
        runner.FirstSeenAt,
        runner.LastSeenAt,
        claim?.Key,
        claim is { } held ? held.Claim.Chatter : runner.Line,
        claim is { } at ? at.Claim.ChatterAt : runner.LineAt,
        runner.State,
        runner.Under,
        runner.MaxRuns,
        runner.MaxSpend,
        runner.UntilAt,
        GoneAfterSeconds);

    /// <summary>What the loop reads back off its own heartbeat: the row's operator half, and nothing else.</summary>
    public static RunnerInstructionDto Instruct(EfHatchRunner runner) =>
        new(runner.State, runner.Under, runner.MaxRuns, runner.MaxSpend, runner.UntilAt);

    /// <summary>
    /// One line at most, trimmed and capped - <see cref="IssueClaimController"/>'s
    /// rule, called on the other kind of heartbeat. Null stays null, because
    /// "no opinion" and "clear it" are different instructions and the empty
    /// string is the second one.
    /// </summary>
    public static string? Normalise(string? line)
    {
        if (line is null) return null;

        var first = line.ReplaceLineEndings("\n").Split('\n')[0].Trim();
        return first.Length > EfHatchRunner.MaxLineLength ? first[..EfHatchRunner.MaxLineLength] : first;
    }
}

File: src/Hatch.Web/apps/hatch/src/lib/runners.ts
/* What a runner row says about itself, apart from the page that draws it - the
   way lib/claim.ts is apart from the badge.

   The server hands over four facts and no verdict: when the runner was last
   heard from, how long silence is allowed to last, what it holds, and what it
   was asked to do. Turning those into one word is this file's whole job, and it
   is here rather than in the page because the ordering between them is a
   judgement worth testing: a runner that has stopped answering is *gone*
   whatever it was asked to do, because "we told it to pause" says nothing about
   whether it is still there to have heard.

   Pure functions taking `now` rather than reading it, so a page judges every
   row against one instant. */

import type { Runner, RunnerPatchRequest } from '../types';

/** The five states a row can be in, in the order they are decided. */
export type RunnerActivity = 'gone' | 'paused' | 'stopping' | 'working' | 'idle';

/**
 * What this runner is doing, in one word.
 *
 * Silence beats intent, which is the one ordering decision here: a paused
 * runner that stopped heartbeating is `gone`, because the pause was a request
 * and the silence is a fact. Below that, what it was asked to do beats what it
 * is holding - a loop finishing its last increment is `stopping`, and that is
 * the thing somebody watching wants to see.
 */
export function runnerActivity(runner: Runner, now: Date): RunnerActivity {
  const beat = Date.parse(runner.lastSeenAt);

  // An unparseable heartbeat is the one case where "look at this" is the honest
  // answer: it is a row whose age cannot be told, not a row that is fine.
  if (Number.isNaN(beat) || now.getTime() - beat > runner.goneAfterSeconds * 1000) return 'gone';

  if (runner.state === 'paused') return 'paused';
  if (runner.state === 'stopping') return 'stopping';

  return runner.claimKey ? 'working' : 'idle';
}

/** The same, as the words on the row. */
export function activityWords(runner: Runner, now: Date): string {
  const activity = runnerActivity(runner, now);
  switch (activity) {
    case 'gone':
      return 'Gone';
    case 'paused':
      return 'Paused';
    case 'stopping':
      return 'Stopping';
    case 'working':
      return `Working ${runner.claimKey}`;
    default:
      return 'Idle';
  }
}

/**
 * Whether this row has controls at all.
 *
 * A `once` runner is one increment that has already read whatever it was going
 * to read; a Pause on it would be a button nothing will ever look at, which is
 * worse than no button.
 */
export const isControllable = (runner: Runner) => runner.kind === 'loop';

/** The bounds as the form holds them: strings, because that is what inputs are
    and what the wire takes. */
export interface RunnerBounds {
  under: string;
  maxRuns: string;
  maxSpend: string;
  /** The wire form, or `''` - what MomentField hands back. */
  untilAt: string;
}

/** What a row's bounds look like in the form that edits them. */
export function boundsOf(runner: Runner): RunnerBounds {
  return {
    under: runner.under ?? '',
    maxRuns: runner.maxRuns === null ? '' : String(runner.maxRuns),
    maxSpend: runner.maxSpend === null ? '' : String(runner.maxSpend),
    untilAt: runner.untilAt ?? '',
  };
}

/** Whether the form is holding anything the row is not. */
export function boundsChanged(draft: RunnerBounds, runner: Runner): boolean {
  const saved = boundsOf(runner);
  return (['under', 'maxRuns', 'maxSpend', 'untilAt'] as const).some((f) => draft[f].trim() !== saved[f].trim());
}

/**
 * Why the form cannot be saved yet, or null.
 *
 * The one rule worth stating on screen rather than discovering as a refusal:
 * `until` is an hour and not a day. The server says the same thing, and this is
 * only what stops somebody typing a date, pressing Save and being told off for
 * a control that offered them the empty time box in the first place.
 */
export function boundsProblem(draft: RunnerBounds): string | null {
  return draft.untilAt.trim() !== '' && !draft.untilAt.includes('T')
    ? 'An until is an hour, not just a day - set a time beside the date.'
    : null;
}

/**
 * The form as a request: every field, trimmed, including the empty ones.
 *
 * Sending all four rather than only what changed is the point of the tri-state:
 * `''` is how a cap comes off, and a field left out would mean "leave it", so a
 * cleared box that was not sent would silently keep the cap it was clearing.
 */
export function boundsRequest(draft: RunnerBounds): RunnerPatchRequest {
  return {
    under: draft.under.trim(),
    maxRuns: draft.maxRuns.trim(),
    maxSpend: draft.maxSpend.trim(),
    untilAt: draft.untilAt.trim(),
  };
}
